mediagrains/utils/h264_parser.py

```diff
@@ -0,0 +1,700 @@
+        self.frame_rate = None
+        self.fixed_frame_rate = None
+        self.sample_aspect_ratio = None
+        self.key_frame = None
+
+
+class H264Parser(object):
+    def __init__(self):
+        self.sps_sets: Dict[int, SPS] = {}
+        self.pps_sets: Dict[int, PPS] = {}
+
+    @classmethod
+    def get_nalu_byte_offsets(cls, frame_data: bytes) -> Iterator[int]:
+        """Returns a generator of NALU offsets in the frame
+
+        The nalu_byte_offsets offsets are at the start code prefix 0x000001.
+        Add 3 to get to the start of the NAL data.
+
+        :param frame_data: frame data bytes
+        :returns: a iterable of byte offsets
+        """
+        frame_bits = BitStream(bytes=frame_data)
+        return (offset//8 for offset in frame_bits.findall('0x000001', bytealigned=True))
+
+    def parse_frame(self, frame_data: bytes) -> Tuple[Optional[int], Optional[Iterable[int]], Optional[FrameInfo]]:
+        """Returns the frame size, unit offsets and frame info if emough data is available.
+
+        Returns None if not enough data is available to detect the end of a frame.
+
+        :param frame_data: frame data bytes
+        :returns: tuple of frame size, NALU byte offsets and frame info, or None for all if there is insufficient data
+        :raises ValueError: if the is an issue with the bitstream
+        :raises MissingParameterSets: a ValueError that is raised if referenced SPS and PPS are unavailable
+        """
+        insufficient_data_return = (None, None, None)
+
+        try:
+            frame_bits = BitStream(bytes=frame_data)
+
+            nalu_offsets = self._get_nalu_offsets(frame_bits)
+            nalu_byte_offsets = []
+
+            offset = next(nalu_offsets, None)
+            if offset is None:
+                # No NALU found
+                return insufficient_data_return
+
+            primary_slice_headers: List[SliceHeader] = []
+            have_start_next_frame = False
+            for next_offset in chain(nalu_offsets, [frame_bits.len]):
+                nal_unit_type = self._parse_nal_unit_type(frame_bits[offset:next_offset])
+
+                if (nal_unit_type == NALUnitTypes.CODED_SLICE_NON_IDR_PICT or
+                        nal_unit_type == NALUnitTypes.CODED_SLICE_DATA_PART_A or
+                        nal_unit_type == NALUnitTypes.CODED_SLICE_IDR_PICT):
+                    rbsp_bits = self._parse_rbsp_bits(frame_bits[offset:next_offset])
+                    slice_header = self._parse_slice_header(rbsp_bits)
+
+                    pps = self.pps_sets[slice_header.pic_parameter_set_id]
+                    sps = self.sps_sets[pps.seq_parameter_set_id]
+
+                    if slice_header.redundant_pic_cnt == 0:
+                        first_slice_header = None
+                        if primary_slice_headers:
+                            first_slice_header = primary_slice_headers[0]
+
+                        if (first_slice_header is not None and
```

Comment:
Might be over-kill but is it worth breaking some of this out into functions? Might help shorten the chain of boolean operators.

Reply:
It's following the list of conditions defined in the H.264 spec. for detecting the start of a new frame. I'l add a comment explaining what it is.